cmd/opm/serve/serve.go

```diff
@@ -59,12 +71,19 @@ will not be reflected in the served content.
 	}
 
 	cmd.Flags().BoolVar(&s.debug, "debug", false, "enable debug logging")
-	cmd.Flags().StringVarP(&s.port, "port", "p", "50051", "port number to serve on")
 	cmd.Flags().StringVarP(&s.terminationLog, "termination-log", "t", "/dev/termination-log", "path to a container termination log file")
+	cmd.Flags().StringVarP(&s.port, "port", "p", "50051", "port number to serve on")
+	cmd.Flags().StringVar(&s.pprofAddr, "pprof-addr", "", "address of startup profiling endpoint")
 	return cmd
 }
 
 func (s *serve) run(ctx context.Context) error {
+	p := newProfilerInterface(s.pprofAddr, s.logger)
+	p.startDefaultEndpoint()
+	if err := p.startCustomEndpoint(); err != nil {
+		return fmt.Errorf("could not start CPU profile: %v", err)
+	}
```

Comment:
These are mostly style/idiomatic nits.

> we are installing a custom handler in the default mux in addition to the pprof handlers we get for just doing the import _ "net/http/pprof".

Yeah, my minor style nit here is that the handlers for the server are configured in several different places, so it's a little hard to build the mental model of "what handlers are configured right now on my server?" I'm also not a huge fan of the "side-effect import" style of stuff because it obfuscates what's happening. Being explicit about the built-in pprof handlers we're running is nicer IMO.

The other nice thing about using `http.Server` is that you have more control over other aspects of it (e.g. server timeouts). I generally use `http.Server` rather than `http.ListenAndServe` because:
1. If I ever need to do more fine-grained configuration, I'm already setup for that.
2. It enables the graceful shutdown that `http.ListenAndServe` precludes.

`http.ListenAndServe` is actually just a [small wrapper](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/net/http/server.go;l=3220-3223) around the snippet I posted above.

> I don't think it matters here. I've run shutdowns where I had a pprof connection to the startup profile endpoint and it continued working just fine

`go tool pprof` doesn't actually hold an open connection to the pprof endpoints as far as I know. The concern here is that if a client is in the middle of downloading a profile when we shutdown, we won't wait for that download to complete. This problem would be more apparent if the profiles we served were large enough that it would be easy to initiate the connection and then kill the server before the download completes. For a debug endpoint like this, I don't think that matters really at all, but I just wanted to bring this up as an FYI for "here's the hardened way to configure/run a webserver using Go's stdlib" just for future reference.

Reply:
Done